### PR TITLE
Add Keycloak 22

### LINKED ISSUE #897: Add support for Keycloak 22
### Current Behavior

Currently a config that worked with Keycloak 21.1.2 cannot be applied against Keycloak 22.0.0.

The following errors are logged in the console:

```
dev-acme-keycloak-provisioning-1  | 2023-07-12 07:21:20.046 DEBUG 1 --- [           main] d.a.k.config.provider.KeycloakProvider   : Attempt failure #17: javax.ws.rs.ProcessingException: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "clientSignatureSymmetricAlgorithms" (class org.keycloak.representations.info.CryptoInfoRepresentation), not marked as ignorable (2 known properties: "supportedKeystoreTypes", "cryptoProvider"])
dev-acme-keycloak-provisioning-1  |  at [Source: (org.jboss.resteasy.specimpl.AbstractBuiltResponse$InputStreamWrapper); line: 1, column: 1173] (through reference chain: org.keycloak.representations.info.ServerInfoRepresentation["cryptoInfo"]->org.keycloak.representations.info.CryptoInfoRepresentation["clientSignatureSymmetricAlgorithms"])
```

### Expected Behavior

Keycloak config cli should be able to apply a configuration against Keycloak 22.0.0

### Steps To Reproduce

_No response_

### Environment

- Keycloak Version: 22.0.0
- keycloak-config-cli Version: 5.6.1-21.0.1
- Java Version: Red_Hat-17.0.7.0.7-2

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by thomasdarimont**
**What this PR does / why we need it**:

This is needed to make keycloak-config-cli compatible with Keycloak 22.

**Which issue this PR fixes** 
Fixes #897

**Special notes for your reviewer**:
- ~~The PR is WIP~~
- Retain compatibility with Keycloak 18, to support current RH-SSO version
- Needed to add `internationalizationEnabled` to some realm import files, since creating users with a locale attribute now requires it (see: [this commit](https://github.com/keycloak/keycloak/commit/d0691b088403149bf78bc671ba1611978ab5ccdb))
- changed the source to use the new `jakarta` packages by default
- adapted older ci builds to use the `javax` packages
- adapted code to be compatible with both resteasy versions 4.7.x and 6.2.x (used by 22.0.0 / quarkus 3)
- Adapted checkstyle to consider jakarta packages as SPECIAL_PACKAGES ordered after STANDARD_JAVA_PACKAGE (since it replaces javax).

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by thomasdarimont** [comment]
@st3v0rr This is still WIP but I think we're almost there.

I think it is a good idea to merge the other danlging PR first to avoid conflicts.
- #865 Add support for managing client-policies
- #861 Allow importing attribute groups in User Declarative Profile

**Comment by jonasvoelcker** [comment]
Hey @thomasdarimont, this one looks very good, thanks :) I guess it is the better solution than mine, but maybe it's worth a look: https://github.com/adorsys/keycloak-config-cli/pull/899

**Comment by thomasdarimont** [comment]
Skipping checkstyle for now to see if we can get past the package order style errors.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -27,10 +27,10 @@ jobs:
       fail-fast: false
       matrix:
         env:
-          - KEYCLOAK_VERSION: 18.0.2
```
Comment (thomasdarimont) [inline .github/workflows/ci.yaml:30]: Btw. I think we should keep Keycloak 18.0.2 if possible as this is AFAIK the last version that is compatible with the latest(?) RH-SSO release.

**Comment by jonasvoelcker** [comment]
@thomasdarimont I tried to adapt most of your changes to my Pull Request and also found that there must be a replacement of jakarta to javax for the existing build pipeline.

This profile might be what you mentioned:

```
        <profile>
            <id>javax</id>
            <build>
                <plugins>
                    <plugin>
                        <groupId>com.google.code.maven-replacer-plugin</groupId>
                        <artifactId>replacer</artifactId>
                        <version>${maven-replacer.version}</version>
                        <executions>
                            <execution>
                                <id>replace-import-javax-by-jakarta</id>
                                <phase>generate-sources</phase>
                                <goals>
                                    <goal>replace</goal>
                                </goals>
                            </execution>
                        </executions>
                        <configuration>
                            <basedir>
                                ${project.basedir}/src
                            </basedir>
                            <includes>
                                <include>**/*.java</include>
                            </includes>
                            <replacements>
                                <replacement>
                                    <token>import jakarta</token>
                                    <value>import javax</value>
                                </replacement>
                            </replacements>
                        </configuration>
                    </plugin>
                </plugins>
            </build>
        </profile>
```

Feel free to use it :)

**Comment by jonasvoelcker** [comment]
I will try to put together some dependencies within the profile after sleeping. ^^

**Comment by thomasdarimont** [comment]
@jonasvoelcker Thanks for your help! I replaced the sed magic with the maven-replacer-plugin and it works now :)

Last task is now to re-enable checkstyle and get rid of the import order errors.

**Comment by thomasdarimont** [comment]
Just build a local docker image based on the latest build and tested it against it with a large list of realm configurations on Keycloak 22.0.0. Works like a charm :)

@st3v0rr I think this PR is now ready for review.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (st3v0rr) [reply .github/workflows/ci.yaml:30]: Is there still a high usage or any other reason not to use older compatible versions of the cli?
I would prefer to keep the maintenance with three versions only.

Comment (thomasdarimont) [reply .github/workflows/ci.yaml:30]: AFAIK every company that uses an up-to-date RH-SSO deployment is effectively on API level 18.0.x.
https://access.redhat.com/articles/2342881

We have a few customers who use RH-SSO, but the vast majority uses Keycloak OSS.

If version 18 is dropped, then new versions of keycloak-config-cli cannot be used with RH-SSO (based on the old Keycloak versions), due too incompatible representations.

I don't know how long RH-SSO stays on the old WF/JBoss EAP based Keycloak, but until then those users can only use older keycloak-cli-versions.

How about:
- we drop support 18 now, and wait for users to complain via reporting issues. 
- Then we could offer to provie "bug-fix" releases for the 5.6.x, which is the last version with Keycloak 18 support.

**Comment by thomasdarimont** [comment]
I'll rebase this PR on current main.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (thomasdarimont) [reply .github/workflows/ci.yaml:30]: alernatively ... we could just try to keep support for 18.0.2 (just until 23.x comes out) to give users more time to upgrade.

**Comment by thomasdarimont** [comment]
@st3v0rr I think the PR is now in good shape. We just did an internal code review that showed no issues.

**Comment by thomasdarimont** [comment]
Just updated my comment.

**Comment by st3v0rr** [comment]
@thomasdarimont I just had 2 more remarks on the readme support text and the maven replacer plugin.
Is it possible to have a look on them?

**Comment by thomasdarimont** [comment]
I don't see your remarks? Did you submit the review?

**Comment by st3v0rr** [comment]
> I don't see your remarks? Did you submit the review?

Sorry my fault, not it's submitted...

File: pom.xml [hunk 2]
```diff
@@ -633,6 +691,41 @@
     </build>
 
     <profiles>
+        <profile>
+            <id>javax</id>
```
[the remark this answers is not in the document]
Comment (thomasdarimont) [reply pom.xml:695]: Yes, see ci.yml in order to replace the "jakarta"  imports  "back" to "javax" in order to be able to build the keycloak-config-cli versions for Keycloak versions < 22.0.0.

**Comment by thomasdarimont** [comment]
Just re-added Keycloak 18.0.2, let's see how this goes.
- Adapted readme and changelog as you suggested.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (thomasdarimont) [reply .github/workflows/ci.yaml:30]: Decision: we keep support for 18.0.x for now.

File: pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (st3v0rr) [reply pom.xml:695]: Ok thanks I didn't get the link. Now it's clear.

**Comment by thomasdarimont** [comment]
@st3v0rr the code coverage step failes due to timeout of external service. I think you have to restart the build again (... happend 2-3 times to me yesterday)

**Comment by st3v0rr** [comment]
> @st3v0rr the code coverage step failes due to timeout of external service. I think you have to restart the build again (... happend 2-3 times to me yesterday)

Now it went through, hopefully it will work with the release :-)

**Pull request merged**